srk/src/transition.ml

```diff
@@ -390,63 +437,142 @@ struct
-    | `Sat -> `Invalid
-    | `Unknown -> `Unknown
-    | `Unsat core ->
-       let core_symbols =
-         List.fold_left (fun core phi ->
-             match Formula.destruct srk phi with
-             | (`Proposition (`App (s, []))) -> Symbol.Set.add s core
-             | _ -> assert false)
-           Symbol.Set.empty
-           core
-       in
-       let (itp, _) =
-         List.fold_right2 (fun tr guard (itp, post) ->
-             let subst sym =
-               match Var.of_symbol sym with
-               | Some var ->
-                  if M.mem var tr.transform then
-                    M.find var tr.transform
-                  else
-                    mk_const srk sym
-               | None -> mk_const srk sym
-             in
-             let post' = substitute_const srk subst post in
-             let reduced_guard =
-               List.filter_map (fun (indicator, guard) ->
-                   if Symbol.Set.mem indicator core_symbols then
-                     Some (mk_not srk guard)
-                   else
-                     None)
-                 guard
-             in
-             let wp =
-               (mk_not srk (mk_or srk (post'::reduced_guard)))
-               |> Quantifier.mbp srk (fun s -> Var.of_symbol s != None)
-               |> mk_not srk
-             in
-             (wp::itp, wp))
-           trs
-           guards
-           ([post], post)
-       in
-       `Valid (List.tl itp)
+    (* gather all symbols into a list, while adding formulas to the solver object *) 
+    let symbols = List.fold_left 
+      (fun symbols (tr, guard) ->
+        let f = to_ss_formula tr guard in 
+          Smt.Solver.add solver [f];
+          (Syntax.symbols f) :: symbols)
+        [] (List.combine trs guards) 
+        |> List.rev
+        |> List.map (fun ss -> Symbol.Set.diff ss indicator_symbols) in 
+    (* subscript the symbols in the `post` formula, as well *)
+    let target = substitute_const srk subscript (mk_not srk post) in
+      Smt.Solver.add solver [target];
+      match Smt.Solver.get_unsat_core_or_model solver indicators with 
+        | `Sat m ->  
+          (sat_callback m symbols ss_inv)
+        | `Unsat core -> (unsat_callback trs post guards core)
+        | `Unknown -> `Unknown 
+
+
+
+  let interpolate ?(solver=Smt.mk_solver C.context) ?(qflia_solver=(Smt.mk_solver ~theory:"QF_LIA" srk)) trs post =
+    Smt.Solver.reset solver;
+    let trs = List.map rename_skolems trs in 
+    interpolate_query ~solver:solver trs post (fun _ _ _ -> `Invalid) @@ interpolate_unsat_core qflia_solver
+
+  let interpolate_or_concrete_model ?(solver=Smt.mk_solver C.context) ?(qflia_solver=(Smt.mk_solver ~theory:"QF_LIA" srk)) trs post = 
+    (* subst_model: rename skolem constants back to their appropriate names using reverse subscript table *)
+    Smt.Solver.reset solver;
+    let trs = List.map rename_skolems trs in 
+    let sat_model model (symbols: Symbol.Set.t list) ss_inv = 
+        let m = 
```

Comment:
subscript_tbl should contain (variable, latest subscript) pairs, so you should be able to construct the model by enumerating over all of them, and adding the (variable, value of the latest subscript within m) binding to the interpretation.